File: Editorial/0014_Longest_Common_Prefix/JavaScript/3_Divide_and_conquer.js
var longestCommonPrefix = function (strs) {
    if (!strs || strs.length < 1) return "";

    let LCP = (str1, str2) => {
        let minLen = Math.min(str1.length, str2.length);
        for (let i = 0; i < minLen; i++) {
            if (str1[i] !== str2[i]) {
                return str1.slice(0, i);
            }
        }
        return str1.slice(0, minLen);
    };

    let divideAndConquer = (strs, l, r) => {
        if (l == r) {
            return strs[l];
        } else {
            let mid = Math.floor((l + r) / 2);
            let lcpLeft = divideAndConquer(strs, l, mid);
            let lcpRight = divideAndConquer(strs, mid + 1, r);
            return LCP(lcpLeft, lcpRight);
        }
    };

    return divideAndConquer(strs, 0, strs.length - 1);
};

File: Editorial/0000-0100/0014_Longest_Common_Prefix/CSharp/3_Divide_and_conquer.cs
public class Solution {
    public string LongestCommonPrefix(string[] strs) {
        if (strs == null || strs.Length == 0)
            return "";
        return LongestCommonPrefix(strs, 0, strs.Length - 1);
    }

    private string LongestCommonPrefix(string[] strs, int l, int r) {
        if (l == r) {
            return strs[l];
        } else {
            int mid = (l + r) / 2;
            var lcpLeft = LongestCommonPrefix(strs, l, mid);
            var lcpRight = LongestCommonPrefix(strs, mid + 1, r);
            return CommonPrefix(lcpLeft, lcpRight);
        }
    }

    private string CommonPrefix(string left, string right) {
        int min = Math.Min(left.Length, right.Length);
        for (int i = 0; i < min; i++) {
            if (left[i] != right[i])
                return left.Substring(0, i);
        }

        return left.Substring(0, min);
    }
}
